File: 0496-next-greater-element-i/0496-next-greater-element-i.cpp
class Solution {
public:
    vector<int> nextGreaterElement(vector<int>& nums1, vector<int>& nums2) {
        vector<int> temp(nums2.size() , -1);
        stack<int> st;
        st.push(nums2[nums2.size() - 1]);
        for(int i = nums2.size() - 2; i >= 0 ; i--){
            while(!st.empty() && nums2[i] >= st.top()) st.pop();
            if(st.empty()) temp[i] = -1;
            else temp[i] = st.top();
            st.push(nums2[i]);
        }
        vector<int> ans(nums1.size());
        for(int i = 0 ; i < nums1.size() ; i++){
            for(int j = 0 ; j < nums2.size() ; j++){
                if(nums1[i] == nums2[j]) ans[i] = temp[j];
            }
        }
        return ans;
    }
};

File: 0496-next-greater-element-i/0496-next-greater-element-i.java
import java.util.Arrays;
class Solution {
    public int[] nextGreaterElement(int[] nums1, int[] nums2) {
        int n = nums1.length;
        int[] ans = new int[n];
        Arrays.fill(ans , -1);
        HashMap<Integer , Integer> mp = new HashMap<>();
        for(int i = 0; i < nums2.length ; i++){
            mp.put(nums2[i] , i);
        }
        for(int i = 0 ; i < n ; i++){
            int idx = mp.get(nums1[i]);
            for(int j = idx + 1; j < nums2.length; j++){
                if(nums2[j] > nums1[i]){
                    ans[i] = nums2[j];
                    break;
                }
            }
        }
        return ans;
    }
}